openebs/mayaserver
# Add '/metrics' endpoint and custom metrics that to be scrapped by prometheus

This commit refers to issue #103, #104 and #105.

1. Why is this change necessary?
* Now [Prometheus](https://prometheus.io) can scrap the metrics from `/latest/volumes`. This metrics 
  contains the details about garbage collector, processes, and memory stats.
* 'gziphandler.GzipHandler(mux)` is creating some issues, if accessing the endpoints from browser. So i 
  it's commented. 

2. How does this change address this issue?
* `promhttp.Handler()` is the handler provided by the [Prometheus](https://prometheus.io/) , which 
   captures how many requests has been made on a particular endpoint with the help of custom labels. 
* Instead of starting the server with argument `promhttp.Handler(mux)`, passing `mux` only resolves the 
  issue.

3. How to verify this change ?
* make new binary by running `make dev`
* now run this new binary `m-apiserver up`
* open `localhost:5656/latest/volumes` and `localhost:5656/metrics` on your browser.
* You will be able to see the error mentioned above for the `/latest/volumes` and metrics collected from 
  maya api service.
e.g. You should see something like this on your browser for `/metrics` endpoint and proper error msg for the other endpoints in the browser.
```# HELP go_gc_duration_seconds A summary of the GC invocation durations.
# TYPE go_gc_duration_seconds summary
go_gc_duration_seconds{quantile="0"} 3.7256e-05
go_gc_duration_seconds{quantile="0.25"} 4.1712e-05
go_gc_duration_seconds{quantile="0.5"} 0.000112669
go_gc_duration_seconds{quantile="0.75"} 0.000148872
go_gc_duration_seconds{quantile="1"} 0.000647127
go_gc_duration_seconds_sum 0.000987636
go_gc_duration_seconds_count 5
# HELP go_goroutines Number of goroutines that currently exist.
# TYPE go_goroutines gauge
go_goroutines 15
# HELP go_info Information about the Go environment.
# TYPE go_info gauge
go_info{version="go1.8.1"} 1
# HELP go_memstats_alloc_bytes Number of bytes allocated and still in use.
# TYPE go_memstats_alloc_bytes gauge
-
-
-
volume_request_duration_seconds_bucket{code="500",method="GET",le="10"} 1
volume_request_duration_seconds_bucket{code="500",method="GET",le="+Inf"} 1
volume_request_duration_seconds_sum{code="500",method="GET"} 0.0001483
volume_request_duration_seconds_count{code="500",method="GET"} 1
# HELP volume_requests_total Total number of /latest/volumes requests.
# TYPE volume_requests_total counter
volume_requests_total{code="500",method="GET"} 1
```
This volume_requests_total captures the no of requests made on `/latest/volumes`. 

4. What side effects does this change have ?

- none
